feat: coroutine nodes

fastcat opened the pull request:
Adds support for writing a node as a coroutine. This allows writing more normal loops and other control flow in a "linear" fashion, instead of having to decompose them down to blackboard state to be rehydrated on the next event.

The accessible way to write coroutines in Go is with iterators, and so that is how this works.

Since `yield` in Go can't return a value, we model the Node coroutine as a "filter": it takes an iterator (generator) as input, providing it the events, and yields results as it iterates over those. The coroutine node must (precisely) alternate between retrieving the next event and yielding its next result.

The example test code works in a simpler `for` loop, but many use cases will may benefit from using `iter.Pull2` so they can fetch single events after taking specific actions, and not have to write the entire node as a single `for` loop, reconstructing where they were in the node logic at each event (which would defeat the point).

The need to update to Go 1.24 and thus update golangci-lint led to a lot of cruft. The actual feature is in [f35a6c0..f11c378](https://github.com/jbcpollak/greenstalk/pull/38/files/f35a6c0f733ca7b920521234cc7f5e1db0feca7a..f11c37858ba5244a21c8f012f56b17cae34f269e)

jbcpollak commented on common/action/async_function_action_test.go:
```diff
@@ -25,7 +25,7 @@ func TestAsyncFunctionAction(t *testing.T) {
 	const error_msg = "error here"
 	asyncFunctionExpectedResultsMap[core.StatusError] = func(ctx context.Context) core.ResultDetails {
-		return core.ErrorResult(fmt.Errorf(error_msg))
+		return core.ErrorResult(errors.New(error_msg))
```
is this new in Go 1.24 or were we just doing it wrong before?

jbcpollak replied:
(nevermind, Copilot explained it)

jbcpollak commented on coro/node_test.go:
```diff
@@ -0,0 +1,65 @@
+	ctx := t.Context()
+
+	ticks, completions := 0, 0
+	tree := coro.Node(
+		func(
+			ctx context.Context,
+			_ core.BaseParams,
+			next iter.Seq2[core.EmptyBlackboard, core.Event],
+		) iter.Seq[core.ResultDetails] {
+			return func(yield func(core.ResultDetails) bool) {
+				for b, e := range next {
+					t.Logf("got tick: %v %T", b, e)
+					ticks++
+					if _, ok := e.(completionEvent); ok {
+						t.Log("ending on completion event")
+						completions++
+						yield(core.SuccessResult())
+						// if we don't break here, `next` will still end
+						break
+					} else {
+						yield(core.RunningResult())
+					}
+				}
+				t.Log("coro node complete")
+			}
+		},
+		core.BaseParams("coro.TestNode"),
+	)
```
What about something like this? 

Changes here:

* Extract the generator from the node, to make it separately testable. Not 100% sure that's useful
* move the context into the iter.Seq2[] so its available on ticks
* removes the blackboard - note this requires state to be passed via "services" in references within params. I'm not sure how we'd handle concurrency and locking of retrieving information within the params though.

```suggestion
         func counter(
	          _ core.BaseParams,
	          next iter.Seq2[ctx context.Context, core.Event],
          ) iter.Seq[core.ResultDetails] {
	          return func(yield func(core.ResultDetails) bool) {
		          for b, e := range next {
			          t.Logf("got tick: %v %T", b, e)
			          ticks++
			          if _, ok := e.(completionEvent); ok {
				          t.Log("ending on completion event")
				          completions++
				          yield(core.SuccessResult())
				          // if we don't break here, `next` will still end
				          break
			          } else {
				          yield(core.RunningResult())
			          }
		          }
		          t.Log("coro node complete")
	          }
          }
          
	tree := coro.Node(
		counter,
		core.BaseParams("coro.TestNode"),
	)
```

fastcat replied:
hoisting out the actual generator to a named thing seems fine. We could also have a variant of `coro.Node` that implicitly generates a `BaseParams` from the func name using some reflection trickery.

jbcpollak reviewed (commented):
I haven't had a chance to look at example_async_retrier_test or example_sequence_test but I looked at the coro.Node and added some thoughts.

Summary thoughts:
* would love to take all the unrelated changes/updates separately, I'm guessing they are interleaved and not a clean commit so I'm not sure the best way to get those changes in separately but they are great.
* This is interesting and I can see how it helps clean up writing nodes - But I can't help wonder rather than adapting from coro -> core, what would be involved in making coro the new core and adapting the other way around, so eventually everything could be coro  "native"
* Rather than the blackboard we use structs passed in via params to track state. These structs can be shared across nodes - since all nodes are is still ticked sequentially I don't think there are any concurrency issues, but I wonder if we'd have to lock them.
* related concurrency question: since everything is ticked synchronously (because of n.fNext() ), is the mutex strictly necessary?

fastcat commented:
> would love to take all the unrelated changes/updates separately, I'm guessing they are interleaved and not a clean commit so I'm not sure the best way to get those changes in separately but they are great.

They largely are in separate commits, so I can cherry-pick out a separate PR with the Go 1.24 + linter updates :+1: 

> what would be involved in making coro the new core

While moving the coro support into the core package makes sense, I think many of the core nodes are easier to write as direct simple nodes. E.g. I would argue that the existing implementation of Sequence is simpler than the coroutine one. The async retrier example is borderline.

Where I see the coroutine model being really helpful is when you need control flow within something that makes sense to keep as a single node. E.g. if you want to gather some input with async validation, that could be much easier to do as a coroutine that does _hand waves_ `for { value = acquire(); if valid(value) { break } else { show_error() } }`. Writing up an example case for that kinda requires a problem domain to use for the example, which I was trying to avoid some here so far.

> related concurrency question: since everything is ticked synchronously (because of n.fNext() ), is the mutex strictly necessary?

Probably not. I was thinking about removing that. Which also points out a missing bit of documentation that the coroutines must not start any goroutines except indirectly via the `yield(running(fn))` pattern, and must not try to retrieve any next values on such goroutines.

----

Looking through the other comments ...

fastcat commented on coro/node.go:
```diff
@@ -0,0 +1,137 @@
+// The `ctx` parameter is only valid during activation, i.e. up until the
+// NodeFunc yields its first result. There is no access to the [context.Context]
+// for later events (TODO).
+type NodeFunc[Blackboard any, P core.Params] func(
+	ctx context.Context,
+	params P,
+	next iter.Seq2[Blackboard, core.Event],
+) iter.Seq[core.ResultDetails]
```
I was thinking of coming back and adding an "args" type to capture the 3-tuple of Context, Blackboard, and Event. But if the blackboard is going away, this could be hard coded to only work with `core.EmptyBlackboard` for now if you want.

fastcat commented on coro/node.go:
```diff
@@ -0,0 +1,137 @@
+
+// Tick implements core.Node.
+func (n *node[Blackboard, P]) Tick(ctx context.Context, b Blackboard, e core.Event) core.ResultDetails {
+	// push b,e to the coroutine
```
Arguably it's a "good" case to panic as it indicates programmer error, but it certainly _could_ recognize that situation and return an Error result instead :+1:

fastcat commented on coro/node.go:
```diff
@@ -0,0 +1,137 @@
+func (n *node[Blackboard, P]) Tick(ctx context.Context, b Blackboard, e core.Event) core.ResultDetails {
+	// push b,e to the coroutine
+	n.nMu.Lock()
+	n.nB, n.nE = &b, e
```
Yeah, or just have a single Args struct to capture all 3 since there's no `iter.Seq3`

fastcat reviewed (commented):
One other thing I want to do is add a "coro.Composite" so that things like the example coro-based Sequence impl can implement Walk correctly.

fastcat commented:
Made #39 that this is now "stacked atop". I can rebase later if you want to de-dupe the commit history, but for now leaving it as merges to make review simpler as the followups come in

fastcat commented:
1. Added the composite stuff and updated the sequence example to "show" it working
2. Added the `Tick` type so all 3 args for each tick are accessible in the coroutine
3. Added some wrapper helpers for the common case of a coroutine that takes neither params nor blackboard
4. Cleaned up naming of things
5. Got rid of the mutex in favor of an `atomic.Pointer` since everything is now in the one `Tick` struct. Still dubious if it needs even that level of protection, but it doesn't hurt for now

An interesting thought around getting rid of the blackboard in favor of the params objects: In the case of coroutine nodes, since the entire node is encapsulated as a single function, any extra params it needs can potentially be captured via closure. That said, if several nodes would use the same params object, then having that params object probably remains cleaner than adding closure layers.

If the blackboard is going away, that seems fine, but also seems like an orthogonal PR?
